In [1]:
import pandas as pd

In [2]:
from statsmodels.formula.api import ols

In [4]:
child = pd.read_csv('child.csv')

In [5]:
child.head()

Aggression  Television  Computer_Games  Sibling_Aggression      Diet  \
0    0.374160    0.172671        0.141907           -0.328216 -0.110303   
1    0.771153   -0.032872        0.709918            0.576837 -0.022990   
2   -0.097728   -0.074460       -0.390141           -0.217184  0.280301   
3    0.015935   -0.004427       -0.408080            0.046223 -0.263479   
4   -0.275385   -0.675239       -0.277778           -0.891045  0.226581   

   Parenting_Style  
0        -0.279034  
1        -1.248167  
2        -0.328063  
3        -1.005119  
4         0.489478

In [7]:
res = ols('Aggression ~ Television', child).fit()

res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             Aggression   R-squared:                       0.025
Model:                            OLS   Adj. R-squared:                  0.024
Method:                 Least Squares   F-statistic:                     17.11
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           3.98e-05
Time:                        10:15:26   Log-Likelihood:                -175.93
No. Observations:                 666   AIC:                             355.9
Df Residuals:                     664   BIC:                             364.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.0005      0.012     -0.041      0.967      -0.025       0.024
Television     0.1634      0.040      4.137      0.000       0.086       0.241
==============================================================================
Omnibus:                       24.471   Durbin-Watson:                   1.931
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               58.038
Skew:                           0.108   Prob(JB):                     2.50e-13
Kurtosis:                       4.430   Cond. No.                         3.23
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# 잔차Residual: 왜 틀렸는지 분석
- 회귀분석의 예측과 실제 값의 차이
- 잔차의 분포에서 주목하는 특성들
    1. 왜도Skewness
    2. 첨도Kurtosis
    3. 등분산성Homoscedasticity

### 잔차의 정규성Nomality
- 잔차가 정규분포에 가까운 성질을 가지고 있는가
- Omnibus / Jarque-Bera
- 둘 다 Prob 가 1 에 가까울수록 정규분포에 가까움

### 1. 왜도Skewness
- 분포의 비대칭성
- 한 방향으로 크게 벗어난다

\ |Negative Skew|Positive Skew
--|--|--
모양|왼쪽으로 길다|오른쪽으로 길다
예시|입지에 따른 부동산 가격|재산
설명|귀신 나오는 집이 있으면|제일 없으면 0, 많으면 무한대

### 2. 첨도Kurtosis
- 분포가 한 점에 몰린 정도
- 정규분포의 첨도 = 3

첨도가 높다|첨도가 낮다
--|--
데이터가 중심에 몰려 있음|데이터가 바깥으로 퍼져 있음

### 3. 등분산성Homoscedasticity: 바람직함
- 모든 범위에서 잔차의 분산이 같다
- 쉽게 말해 어떤 x 에서든 비슷한 정도로 y 를 맞출 수 있다
- Dubin-Watson 통계량이 1~2 정도면 적당하다

### 조건수 Condition Number
- 입력의 변화에 따른 출력의 변화를 나타내는 수
- 얼마나 입력에 민감한지 나타내는 수 
    
        ex. 수도꼭지를 조금만 돌려도 찬물 나오는 것
- 조건수가 크면 데이터가 조금만 달라져도 결과에 큰 차이
- 통상 30 이하

### **다중공선성Multicollinearity
- 독립변수들이 서로 예측가능할 경우(계수 추정이 불안정하면)
- 조건수가 커진다
- 데이터나 변수의 변화에 따라 추정된 계수가 크게 달라진다

### 다중공선성Multicollinearity 예시

In [10]:
child['TV2'] = child['Television']

In [12]:
ols('Aggression ~ Television + TV2', child).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             Aggression   R-squared:                       0.025
Model:                            OLS   Adj. R-squared:                  0.024
Method:                 Least Squares   F-statistic:                     17.11
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           3.98e-05
Time:                        10:45:11   Log-Likelihood:                -175.93
No. Observations:                 666   AIC:                             355.9
Df Residuals:                     664   BIC:                             364.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.0005      0.012     -0.041      0.967      -0.025       0.024
Television     0.0817      0.020      4.137      0.000       0.043       0.120
TV2            0.0817      0.020      4.137      0.000       0.043       0.120
==============================================================================
Omnibus:                       24.471   Durbin-Watson:                   1.931
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               58.038
Skew:                           0.108   Prob(JB):                     2.50e-13
Kurtosis:                       4.430   Cond. No.                     2.35e+16
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.21e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

# 변수 선택 Variable Selection

### 교차검증 Cross Validation
- 데이터를 무작위로 두 세트로 나눔
- 한 세트에서 추정 -> 다른 세트에서 검증

        한 세트에서만 잘 나오면 과적합
- 위 과정을 반복한다

### K-fold Cross Validation
- 데이터를 k 개로 나누어 CV 를 k 번 하는 방법

>ex. k = 3 
>- 1, 2 번 데이터로 추정 -> 3 번 데이터로 검증
>- 2, 3 번 데이터로 추정 -> 1 번 데이터로 검증
>- 1, 3 번 데이터로 추정 -> 2 번 데이터로 검증

        Bootstrapping 과 비슷한 원리

# 과적합 막기
1. 변수 줄이기
2. 계수 줄이기: 정규화
    
    - |계수|↓: Lasso Regression
    - 계수²↓: Ridge Regression
    - 둘 다 ↓: Elastic Net
    
    
3. 곡선 -> 직선: 선형회귀분석
    
        데이터가 많아야 Cross Validation 할 수 있다
        Cross Validation 하면 시간이 더 많이 걸린다

### 정규화 Reaulatrization:과적합Overfitting을 피하기 위해 계수를 줄이기
- 가능한 계수를 작게 추정하는 방법


- 과적합 overfitting: 모형의 개수가 주어진 데이터에 지나치게 의존하여 추정되는 것
    - 변수가 많을수록, 계수가 클수록 과적합의 위험이 크다
    - 만약 계수가 0 이라면 변수를 추가하지 않은 것과 같다
    
\ |복잡한 모형|단순한 모형
--|--|--
모형의 모양|곡선|직선
변수|많다|적다
계수|크다|작다

    복잡한 모형을 만들수록 과적합될 가능성이 높다

### 1. 라쏘 회귀분석 Lasso Regression
- 계수(W)의 절대값을 최소화
        MSE + λ∑|W|
- λ lambda: 클수록 계수를 최소화하는데 더 큰 비중
- 회귀계수를 0으로 만드는 경향

### 2. 릿지 회귀분석 Ridge Regression
- 계수의 제곱을 최소화
        MSE + λ∑W²
- Lasso 보다 대체로 나음, 회귀계수가 0 이 되지는 않는다
- 딥러닝에서 '가중치 감쇠 Weight Decay' 라고 부른다

### 3. 엘라스틱 넷 Elastic Net
- Lasso + Ridge
         MSE + λ(a∑|W| + (1 - a)∑W²)
- a = 1: Lasso
- a = 0: Ridge

### 하이퍼파라미터Hyperparameter
- 모형의 특성을 결정하지만 데이터로부터 학습되지 않는 값
- 엘라스틱 넷Elastic Net 에서 λ 와 a
- CV 를 통해 결정한다

# 정규화Regularization

In [15]:
from sklearn.linear_model import LinearRegression, ElasticNetCV

In [16]:
x = child.iloc[:, 1:6] # 1 번열부터 6 번열까지를 x
y = child['Aggression'] # Aggression 열을 y

# 선형모형Linear

In [18]:
m1 = LinearRegression()

In [19]:
m1.fit(x, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [20]:
m1.coef_

array([ 0.03291628,  0.14216117,  0.08168417, -0.10905412,  0.0566481 ])

# 엘라스틱넷Elastic Net + 교차검증Cross Validation
l1_ratio : 교안에서 a, Lasso(l1), Ridge(l2)

l1_ratio[.1, .5, 1]: 3 가지의 선택지 줌

In [21]:
m2 = ElasticNetCV(l1_ratio=[.1, .5, 1], cv=3) # 3-fold Cross Validation
m2.fit(x, y)
m2.coef_

array([ 0.02572348,  0.12471677,  0.0687051 , -0.08714782,  0.05652955])

### 위의 선형모형으로 나온 계수보다 조금씩 작아졌다

In [25]:
m2.l1_ratio_ # l1_ratio 가 0.1 일 때 가장 과적합이 덜 되더라

0.1

In [26]:
m2.alpha_ # λ 에 해당하는 부분, 0.0088 일 때 가장 과적합이 덜 되더라

0.008882602491363027

# Overfitting: 성급한 일반화의 오류
    ex. 기업 채용
    - 변수(요구 스펙)가 많아지고, 
    - 데이터 사이즈는 작아지고
    
    → 과연 안경 낀 사람들이 일을 잘한다(Overfitting 되기 쉽다)

# 더미 코딩 Dummy Coding
- 독립변수에 이산형 변수가 있을 경우(ex. 짜장/짬뽕/볶음밥)
- 기준이 되는 값을 정함(ex. 짜장)
- 나머지 값을 새로운 변수로 추가(ex. 짬뽕, 볶음밥)
- 해당되는 변수의 값을 1, 나머지는 0 으로 설정한다

점심|점심_짬뽕|점심_볶음밥
-|-|-
짜장|0|0
짬뽕|1|0
볶음밥|0|1

    → 짜장 칼럼을 추가하지 않는 이유: 다중공선성Multicollinearity 
    → 짬뽕 0 볶음밥 0 이면 자동으로 짜장이다

In [31]:
hsb = pd.read_csv('hsb2.csv')
hsb.head()

# SES Socio Economic Status

id  female  race  ses  schtyp  prog  read  write  math  science  socst
0   70       0     4    1       1     1    57     52    41       47     57
1  121       1     4    2       1     3    68     59    53       63     61
2   86       0     4    3       1     1    44     33    54       58     31
3  141       0     4    3       1     3    63     44    47       53     56
4  172       0     4    2       1     2    47     52    57       53     61

In [32]:
# C(변수명): Category 니까 알아서 Dummy Coding 하라
ols('read ~ C(race)', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.084
Model:                            OLS   Adj. R-squared:                  0.070
Method:                 Least Squares   F-statistic:                     5.964
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           0.000654
Time:                        14:41:20   Log-Likelihood:                -740.06
No. Observations:                 200   AIC:                             1488.
Df Residuals:                     196   BIC:                             1501.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       46.6667      2.019     23.117      0.000      42.685      50.648
C(race)[T.2]     5.2424      3.601      1.456      0.147      -1.859      12.344
C(race)[T.3]     0.1333      2.994      0.045      0.965      -5.772       6.038
C(race)[T.4]     7.2575      2.179      3.330      0.001       2.959      11.556
==============================================================================
Omnibus:                        3.544   Durbin-Watson:                   1.966
Prob(Omnibus):                  0.170   Jarque-Bera (JB):                2.702
Skew:                           0.147   Prob(JB):                        0.259
Kurtosis:                       2.512   Cond. No.                         8.25
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- 인종1 과 인종2 
- 인종1 과 인종3
- 인종1 과 인종4
를 각각 비교하고 있다


여기서는 인종2 와 인종4 를 비교할 수는 없다

### pd.get_dummies: 수동으로 더미 코딩

In [41]:
race = pd.get_dummies(hsb['race'], prefix='race')

### 인종4 를 기준으로 삼기

In [49]:
race.iloc[:, 0:3].head()

race_1  race_2  race_3
0       0       0       0
1       0       0       0
2       0       0       0
3       0       0       0
4       0       0       0

In [47]:
new = pd.concat([hsb, race], axis=1)
new.head()

id  female  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   race_1  race_2  race_3  race_4  
0       0       0       0       1  
1       0       0       0       1  
2       0       0       0       1  
3       0       0       0       1  
4       0       0       0       1

In [48]:
ols('read ~ race_1 + race_2 + race_3', new).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.084
Model:                            OLS   Adj. R-squared:                  0.070
Method:                 Least Squares   F-statistic:                     5.964
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           0.000654
Time:                        14:54:32   Log-Likelihood:                -740.06
No. Observations:                 200   AIC:                             1488.
Df Residuals:                     196   BIC:                             1501.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     53.9241      0.821     65.658      0.000      52.304      55.544
race_1        -7.2575      2.179     -3.330      0.001     -11.556      -2.959
race_2        -2.0150      3.093     -0.652      0.515      -8.115       4.085
race_3        -7.1241      2.359     -3.020      0.003     -11.776      -2.472
==============================================================================
Omnibus:                        3.544   Durbin-Watson:                   1.966
Prob(Omnibus):                  0.170   Jarque-Bera (JB):                2.702
Skew:                           0.147   Prob(JB):                        0.259
Kurtosis:                       2.512   Cond. No.                         4.56
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### → 기준 컬럼을 바꿔도 결과는 같다

# .replace 로 컬럼값 바꾸기

In [54]:
hsb['race2'] = hsb['race'].replace({1: 'Caucasian', 2:'Mongoloid', 3:'Negroid', 4:'Australoid'})
hsb.head()

id  female  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

        race2  
0  Australoid  
1  Australoid  
2  Australoid  
3  Australoid  
4  Australoid

In [55]:
ols('read~ C(race2)', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.084
Model:                            OLS   Adj. R-squared:                  0.070
Method:                 Least Squares   F-statistic:                     5.964
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           0.000654
Time:                        15:09:49   Log-Likelihood:                -740.06
No. Observations:                 200   AIC:                             1488.
Df Residuals:                     196   BIC:                             1501.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                53.9241      0.821     65.658      0.000      52.304      55.544
C(race2)[T.Caucasian]    -7.2575      2.179     -3.330      0.001     -11.556      -2.959
C(race2)[T.Mongoloid]    -2.0150      3.093     -0.652      0.515      -8.115       4.085
C(race2)[T.Negroid]      -7.1241      2.359     -3.020      0.003     -11.776      -2.472
==============================================================================
Omnibus:                        3.544   Durbin-Watson:                   1.966
Prob(Omnibus):                  0.170   Jarque-Bera (JB):                2.702
Skew:                           0.147   Prob(JB):                        0.259
Kurtosis:                       2.512   Cond. No.                         4.56
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# 상호작용Interaction
- 두 독립변수의 곱으로 이뤄진 항(xm)
        y = x + m + xm
- 간단히 하기 위해 m 은 0 또는 1 만 갖는 범주형 변수라 하자

### 상호작용이 없는 경우
        y = x + m
- m 에 따라 x 의 절편이 바뀌는 것으로 해석한다

### 상호작용이 있는 경우(1)
         y = x + xm
         y = (1 + m)x
- m 에 따라 x 의 기울기가 바뀌는 것으로 해석

### 상호작용이 있는 경우(2)
        y = x + m + xm
        y = (1 + m)x + m
- m 에 따라 x 의 절편과 기울기가 바뀌는 것으로 해석

### 각 독립변수 사이에 상호작용이 있는지 없는지 판단하기
둘 다 돌려보고 신뢰구간 확인

# 사회경제적수준SES 이 읽기능력에 영향을 주는지

In [58]:
ols('read ~ ses', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.086
Model:                            OLS   Adj. R-squared:                  0.081
Method:                 Least Squares   F-statistic:                     18.64
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           2.49e-05
Time:                        15:46:35   Log-Likelihood:                -739.80
No. Observations:                 200   AIC:                             1484.
Df Residuals:                     198   BIC:                             1490.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     43.6972      2.095     20.858      0.000      39.566      47.829
ses            4.1522      0.962      4.317      0.000       2.256       6.049
==============================================================================
Omnibus:                        6.283   Durbin-Watson:                   1.816
Prob(Omnibus):                  0.043   Jarque-Bera (JB):                3.593
Skew:                           0.104   Prob(JB):                        0.166
Kurtosis:                       2.377   Cond. No.                         7.82
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [63]:
# School Type 이 1/2 로 나타나있어서 0/1로 바꿔준다
hsb['schtyp2'] = hsb['schtyp'] - 1

In [60]:
ols('read ~ ses + schtyp2', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.088
Model:                            OLS   Adj. R-squared:                  0.079
Method:                 Least Squares   F-statistic:                     9.529
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           0.000112
Time:                        15:48:47   Log-Likelihood:                -739.57
No. Observations:                 200   AIC:                             1485.
Df Residuals:                     197   BIC:                             1495.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     43.6743      2.098     20.816      0.000      39.537      47.812
ses            4.0613      0.972      4.178      0.000       2.144       5.979
schtyp2        1.3109      1.916      0.684      0.495      -2.467       5.089
==============================================================================
Omnibus:                        6.257   Durbin-Watson:                   1.804
Prob(Omnibus):                  0.044   Jarque-Bera (JB):                3.633
Skew:                           0.115   Prob(JB):                        0.163
Kurtosis:                       2.381   Cond. No.                         7.85
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [64]:
ols('read ~ ses + ses:schtyp2', hsb).fit().summary() # ses:schtyp2 곱한 것: 상호작용

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.078
Method:                 Least Squares   F-statistic:                     9.376
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           0.000129
Time:                        15:50:17   Log-Likelihood:                -739.71
No. Observations:                 200   AIC:                             1485.
Df Residuals:                     197   BIC:                             1495.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      43.7565      2.104     20.799      0.000      39.608      47.905
ses             4.0600      0.987      4.114      0.000       2.114       6.006
ses:schtyp2     0.3568      0.822      0.434      0.665      -1.265       1.979
==============================================================================
Omnibus:                        6.259   Durbin-Watson:                   1.810
Prob(Omnibus):                  0.044   Jarque-Bera (JB):                3.603
Skew:                           0.108   Prob(JB):                        0.165
Kurtosis:                       2.379   Cond. No.                         7.99
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 해석
    read = (4.06 + 0.35 * School Type) * SES
    School Type = 0, read = 4.06 * SES
    School Type = 1, read = 4.41 * SES

    신뢰구간을 고려하지 않으면,
    School Type 1 일 때, 사회경제적 수준이 성적에 미치는 영향이 더 크다

    → 신뢰구간이 - ~ + 에 걸쳐 있기 때문에 영향이 없다

→ School Type 은 SES 가 읽기 능력에 미치는 영향을 간접적으로 조절한다

    ex. 업무탈진(번아웃)
    업무량 -> 직무탈진
    권한↑이 간접적 영향

    → 권한이 많으면 직무탈진이 안온다

### cf. 통계적 통제
- <사회경제적 수준>과 <학교타입>이 상관관계가 있으니 통계적으로 통제해준다
- <사회경제적 수준>과 <학교타입> 두가지 다 독립변수로 넣어주면 통계적으로 통제된다

In [67]:
ols('read ~ ses + schtyp2', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.088
Model:                            OLS   Adj. R-squared:                  0.079
Method:                 Least Squares   F-statistic:                     9.529
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           0.000112
Time:                        16:03:29   Log-Likelihood:                -739.57
No. Observations:                 200   AIC:                             1485.
Df Residuals:                     197   BIC:                             1495.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     43.6743      2.098     20.816      0.000      39.537      47.812
ses            4.0613      0.972      4.178      0.000       2.144       5.979
schtyp2        1.3109      1.916      0.684      0.495      -2.467       5.089
==============================================================================
Omnibus:                        6.257   Durbin-Watson:                   1.804
Prob(Omnibus):                  0.044   Jarque-Bera (JB):                3.633
Skew:                           0.115   Prob(JB):                        0.163
Kurtosis:                       2.381   Cond. No.                         7.85
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# Interaction 예시

In [68]:
ols('write ~ ses + ses:female', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  write   R-squared:                       0.116
Model:                            OLS   Adj. R-squared:                  0.107
Method:                 Least Squares   F-statistic:                     12.90
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           5.42e-06
Time:                        16:11:01   Log-Likelihood:                -720.78
No. Observations:                 200   AIC:                             1448.
Df Residuals:                     197   BIC:                             1457.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     46.6663      1.914     24.378      0.000      42.891      50.441
ses            1.7444      0.909      1.918      0.057      -0.049       3.538
ses:female     2.3479      0.583      4.027      0.000       1.198       3.498
==============================================================================
Omnibus:                       11.857   Durbin-Watson:                   1.792
Prob(Omnibus):                  0.003   Jarque-Bera (JB):                7.477
Skew:                          -0.324   Prob(JB):                       0.0238
Kurtosis:                       2.309   Cond. No.                         8.80
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 해석
    y = 1.7*SES + 2.3*SES*female
    y = (1.7 + 2.3female)*SES
    female = 0 남자, y = 1.7 * SES
    female = 1 여자, y = 4.0 * SES
    
    → 여자는 남자에 비해 사회경제적 수준이 성적에 영향을 많이 준다

In [69]:
# 'ses * female' == ses + ses:female + female'
ols('write ~ ses + ses:female + female', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  write   R-squared:                       0.124
Model:                            OLS   Adj. R-squared:                  0.111
Method:                 Least Squares   F-statistic:                     9.256
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           9.38e-06
Time:                        16:17:36   Log-Likelihood:                -719.84
No. Observations:                 200   AIC:                             1448.
Df Residuals:                     196   BIC:                             1461.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     43.3109      3.120     13.881      0.000      37.157      49.464
ses            3.1618      1.382      2.288      0.023       0.437       5.887
ses:female     0.0388      1.795      0.022      0.983      -3.501       3.578
female         5.3668      3.946      1.360      0.175      -2.415      13.149
==============================================================================
Omnibus:                       11.529   Durbin-Watson:                   1.789
Prob(Omnibus):                  0.003   Jarque-Bera (JB):                7.780
Skew:                          -0.349   Prob(JB):                       0.0205
Kurtosis:                       2.332   Cond. No.                         22.5
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 해석 
    w = 3.16*SES + 0.03*SES*Female + 5.36*Female
- SES 중심 해석
    w = (3.16 + 0.03*Female)*SES + 5.36*Female
- Female 중심 해석
    w = (0.03*SES + 5.36)*Female + 3.16*SES
    
        여자 w = 3.19*SES + 5.36
        남자 w = 3.16*SES

# 쓰기능력에 미치는 읽기능력과 성별의 상호작용

In [70]:
ols('write ~ read + read:female', hsb).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  write   R-squared:                       0.429
Model:                            OLS   Adj. R-squared:                  0.423
Method:                 Least Squares   F-statistic:                     73.90
Date:                Tue, 17 Sep 2019   Prob (F-statistic):           1.13e-24
Time:                        16:24:37   Log-Likelihood:                -677.12
No. Observations:                 200   AIC:                             1360.
Df Residuals:                     197   BIC:                             1370.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      23.2004      2.654      8.741      0.000      17.966      28.434
read            0.5144      0.050     10.218      0.000       0.415       0.614
read:female     0.0961      0.019      5.000      0.000       0.058       0.134
==============================================================================
Omnibus:                        7.827   Durbin-Watson:                   1.974
Prob(Omnibus):                  0.020   Jarque-Bera (JB):                4.749
Skew:                          -0.201   Prob(JB):                       0.0931
Kurtosis:                       2.361   Cond. No.                         322.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### 상호작용이 없는 경우
    y = ax + bm + c
    m = 0, y = ax + c
    m = 1, y = ax + (b + c)
#### 상호작용이 있는 경우(1)
    y = ax + bm:x + c
    m = 0, y = ax + c
    m = 1, y = (a + b)x + c
#### 상호작용이 있는 경우(2)
    y = ax + bx:m + cm + d
    m = 0, y = ax + d 
    m = 1, y = (a + b)x + (c + d)